Default to SHA-256 for cookie signing

jg23497 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

Fixes #54926. This pull request has been created so that we use the SHA256 algorithm by default for signed cookies, moving away from SHA1.

Unless overriding `config.action_dispatch.signed_cookie_digest`, Rails currently defaults to SHA1 for the generation of the signed cookie HMAC.

As the [issue](https://github.com/rails/rails/issues/54926) implies, use of the SHA1 algorithm is [no longer recommended](https://www.nist.gov/news-events/news/2022/12/nist-retires-sha-1-cryptographic-algorithm), even if there is currently no known attack vector specifically for HMACs. Rails is already migrating to SHA256 in related areas.

### Detail

This pull request changes the default value of `signed_cookie_digest`.

**Note**: Rails applications relying on the previous SHA1 default will no longer read these SHA1-signed cookies unless a cookie rotation is configured:

```
Rails.application.config.action_dispatch.cookies_rotations.tap do |cookies|
  cookies.rotate :signed, digest: "SHA1"
end
```

This rotation allows the upgrading applications to read SHA1 signed cookies created prior to this change without breaking sessions or user data.

The rotation can be removed once no clients are sending cookies signed using the previous digest (e.g. the session lifetimes or cookie expiration times have passed). 

### Additional information

#### Example

Prior to this change, imagine we have the following action method where we set a signed cookie:

```ruby
def index
  cookies.signed[:foo] = { value: 'bar' }
end
```

Cookie value:

```
eyJfcmFpbHMiOnsibWVzc2FnZSI6IkltSmhjaUk9IiwiZXhwIjpudWxsLCJwdXIiOiJjb29raWUuZm9vIn19--a1d2517adb63df049f06fe652040f18b10555ebd
```

Base64 decoding the first part, we see: `{"_rails":{"message":"ImJhciI=","exp":null,"pur":"cookie.foo"}}`. The SHA1 HMAC follows: `a1d2517adb63df049f06fe652040f18b10555ebd`.

After applying this PR's change, restarting the server and attempting to read the cookie using another action method:

```ruby
def foo
  message = cookies.signed[:foo]
  render plain: "#{message || "Unable to read cookie"}\n"
end
```

We see `Unable to read cookie` because the client presented a SHA1-signed cookie, but we're now using SHA256.

We can apply the rotation to gracefully handle this scenario:

```ruby
# config/initializers/cookie_rotations.rb
Rails.application.config.action_dispatch.cookies_rotations.tap do |cookies|
  cookies.rotate :signed, digest: "SHA1"
end
```

The action method now renders `bar` and the cookie's value is updated, including the new SHA256 HMAC:

```
eyJfcmFpbHMiOnsibWVzc2FnZSI6IkltSmhjaUk9IiwiZXhwIjpudWxsLCJwdXIiOiJjb29raWUuZm9vIn19--80d33014a7c363278ff169d809a8740399f0fbaec9d6e944cd6dfed89581f83c
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

jg23497 committed:
Default to SHA256 for the signed cookie digest.

Fixes #54926. Although no known attack vector exists for SHA1 message
authentication specifically, the algorithm should be deprecated due to
its weakness. Rails is also migrating to SHA256 more generally and in
related areas.

jg23497 commented:
Reviewers, this is my first proposed contribution to rails/rails and so I am not sure whether this change warrants a CHANGELOG entry — please let me know. My primary concern was ensuring that the note on cookie rotations is included in the upcoming release notes.

Earlopain commented on guides/source/configuring.md:
```diff
@@ -2067,7 +2067,7 @@ Sets the cipher to be used for encrypted cookies. This defaults to
 #### `config.action_dispatch.signed_cookie_digest`
 
-Sets the digest to be used for signed cookies. This defaults to `"SHA1"`.
+Sets the digest to be used for signed cookies. This defaults to `"SHA256"`.
```
The default should depend on `load_defaults` so that upgrading/legacy apps continue to work. Previous PRs for similar things should give a good template you can take inspiration from: https://github.com/rails/rails/pull/41043

This probably also needs some update https://github.com/rails/rails/blob/96a2f03a2a82e91ab4438b625ba47bd8ed917b45/guides/source/security.md#L333-L347

There are also some tests like https://github.com/rails/rails/blob/96a2f03a2a82e91ab4438b625ba47bd8ed917b45/actionpack/test/dispatch/cookies_test.rb#L630-L631 that don't make much sense anymore when the default is already SHA256.

jg23497 replied:
Thanks for the pointers!

I have done as you suggested with `load_defaults`. I notice that the `signed_cookie_digest` configuration was [added in Rails 5.2](https://github.com/rails/rails/pull/29716) and so have added the previous "SHA1" default to the method's "5.2" switch clause.

I will revisit cookies_test.rb soon to check the redundant tests, hopefully tomorrow.

jg23497 replied:
I have updated the relevant unit tests.

I can also confirm that `config.load_defaults 8.1` produces a SHA256 signed cookie digest and `config.load_defaults 8.0` produces SHA1.

Earlopain reviewed (commented):
👍 I have one concern which I'm fairly sure will need to be adressed. A changelog entry seems fine, but you should also add this to the upgrading rails guide, including the cookie rotator for the (probably many) people that need it.

Earlopain commented:
If you rebase this PR against main, most (all?) CI failures should be resolved.

KelseyDH commented:
I'm upgrading an application to Rails 7.1 right now and was about to delete our SHA256 cookie rotator.  I assume with this bug that I should I also set in `application.rb`:

```ruby
  config.action_dispatch.signed_cookie_digest = "SHA256"
```

To avoid this problem, until a bug fix gets merged into a new Rails release?

jg23497 commented:
> I'm upgrading an application to Rails 7.1 right now and was about to delete our [SHA1 to SHA256 cookie rotator from the Rails upgrade guides](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#key-generator-digest-class-change-requires-a-cookie-rotator). (As part of enabling `config.active_record.encryption.support_sha1_for_non_deterministic_encryption = false`)
> 
> I assume with this bug that I should also set:
> 
> ```ruby
>   # config/application.rb
>   
>   config.action_dispatch.signed_cookie_digest = "SHA256"
> ```
> 
> to avoid this problem, until this bug fix gets merged into a new Rails release?

I would agree that SHA256 is preferable here given the weakness of SHA1, which is generally deprecated. However, I understand that there is no known attack vector for SHA1 HMACs specifically and so no need for immediate concern.

jg23497 commented:
> I am by no means an expert on hashing algorithms but I have done some reading and have found that the SHA-2 can be vulnerable to length extension attacks but SHA-3 is not. Should the hashing algorithm be set to SHA-3 by default ?

Good point that we may want to consider a different default. SHA256 would at least align with the other Rails defaults (e.g. ActiveRecord's hash digest class) and so would be a good contender. 

As far as I understand, length extension attacks do not apply to SHA256 HMACs because the attacker would also need to have knowledge of the key — this is not a regular hash that considers only the message/value itself.

jg23497 committed:
Restore "SHA1" fallback in signed_cookie_digest method.

Allow the framework configuration for Rails 8.1 onwards to control
the signed cookie digest. "SHA1" remains as a fallback and default
for previous versions.

jg23497 commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -359,6 +360,10 @@ def load_defaults(target_version)
           if respond_to?(:action_controller)
             action_controller.escape_json_responses = false
           end
+
+          if respond_to?(:action_dispatch)
+            action_dispatch.signed_cookie_digest = "SHA256"
+          end
```
Hi @p8. That approach will simplify the changes here -  I have done as you suggest.

There is no functional difference as you point out, but my slight concern with this approach is that the `signed_cookie_digest` method now/still implies that "SHA1" is the _current_ default signed cookie digest. However, that is no longer true now that SHA256 is the default and that SHA1 is deprecated, becoming part of the legacy configuration. In other words, we're falling back to a digest that we actually no longer recommend.

Happy to take guidance here, and the above may not be a concern, but sharing just in case.

p8 replied:
We try to not break applications when they get upgraded. The `frameworks_defaults` file is used to allow developers to upgrade to breaking changes on their own choosing. So if someone upgrades Rails but does not yet load the new defaults, nothing should change.

Adding a setting to older Rails version defaults could break things unexpectedly.
For example, someone might define the cookie digest before loading the defaults:
```ruby
config.action_dispatch.signed_cookie_digest = "SHA256"
config.load_defaults 8.0 # this would set the digest back to SHA1 if the defaults set it to SHA1
```
